### PR TITLE
New way to limit resource usage for ota-client during ota-update.

### PR DESCRIPTION
**Pull request by Bodong-Yang**
# Introduction
Previously, we directly cap the `max_workers` of threadpool to `7` for all operations(copy, link, download).
That is not so flexible and not so compatible to different machine. This PR changes the limitation mechanism from capping `max_workers` of threadpool, to using `Semaphore` to limit the total number concurrent tasks(copy, link, download), and apply another Semaphore to specially limit the on-going downloading tasks.

# Changes
1. Apply semaphore(set to 128 currently) to all operations(copy, link, download).
2. Apply semaphore(set to 16 currently) to download operation.
3. Above settings and some other settings are configured at `configs.py`.
4. Use dataclass for all Inf class.

# TODO
- [x] regression test on VM.

### CODE REVIEW

File: app/ota_client.py [hunk 1]
```diff
@@ -237,98 +242,106 @@ def download(
         return error_count
 
 
+def de_escape(s: str) -> str:
+    return s.replace(r"'\''", r"'")
+
+
+@dataclass
```
Comment (obi-t4) [inline app/ota_client.py:249]: I think dataclass decorator is not required since __init__ function is implemented.
The dataclass with __init__ function seems to be complicated for me.
https://stackoverflow.com/questions/54927763/control-initialize-order-when-python-dataclass-inheriting-a-class
I'd like to revert these changes.

File: app/ota_client.py [hunk 2]
```diff
@@ -556,10 +569,12 @@ def __init__(
         self,
         store: OtaClientStatistics,
         *,
+        max_concurrency_tasks: int,
         interval=1,
     ) -> None:
         self.abort_event = threading.Event()
         self.last_error = None
+        self.se = threading.Semaphore(max_concurrency_tasks)
```
Comment (obi-t4) [inline app/ota_client.py:577]: In the previous review, semaphore member is removed from this class, and semaphore is used as callback argument.
Is this change intended?

**Review by obi-t4 — COMMENTED** [review]
I left some comments, please have a check.

File: app/ota_client.py (hunk 1 above)
Comment (Bodong-Yang) [reply app/ota_client.py:249]: Oh I don't use the dataclass generated __init__ method, but write one, so the problem stated in your link doesn't exist in our case. Actually the behavior is the same as before, and we get the benefits from the dataclass.

File: app/ota_client.py (hunk 2 above)
Comment (Bodong-Yang) [reply app/ota_client.py:577]: Yes, it is intended.
Previously this semaphore is created outside `_CreateRegularStatsCollector`, and used as a parameter for callback, and directly used in the threadpool for loop. 
Now I change to initialize the semaphore in the `_CreateRegularStatsCollector` class, and the this semaphore is accessed as `_CreateRegularStatsCollector` instance's property in the for loop.

What I want to achieve is to let the Collector instance take the whole control of acquire/release semaphore. (Oh for better understanding, maybe I should wrap `_collector.se.acquire()` in line 1107 into a method?)

Comment (Bodong-Yang) [reply app/ota_client.py:577]: Ah I may also let the collector control the exception detection.
https://github.com/tier4/ota-client/blob/00ca4b70bb009ee2a67817ac7c4da12626676c59/app/ota_client.py#L1119-L1132

File: app/ota_client.py (hunk 1 above)
Comment (obi-t4) [reply app/ota_client.py:249]: > we get the benefits from the dataclass.

In this change, what is the benefits from the dataclass?

File: app/ota_client.py [hunk 3]
```diff
@@ -237,98 +242,106 @@ def download(
...
-        del self._left
 
+@dataclass
+class RegularInf:
+    mode: int
+    uid: int
+    gid: int
+    nlink: int
+    sha256hash: str
+    size: int
```
Comment (obi-t4) [inline app/ota_client.py:322]:
```suggestion
    size: Optional[int]
```

File: app/ota_client.py [hunk 4]
```diff
@@ -1170,19 +1183,24 @@ def _create_regular_file(
...
                 else:
-                    processed.errors = downloader(
-                        reginf.path,
-                        dst,
-                        reginf.sha256hash,
-                    )
-                    processed.op = "download"
+                    # limit the concurrent downloading tasks
+                    with download_limiter:
+                        if True:
```
Comment (obi-t4) [inline app/ota_client.py:1188]: Is this not required?

File: app/ota_client.py (hunk 2 above)
Comment (obi-t4) [reply app/ota_client.py:577]: > should wrap _collector.se.acquire() in line 1107 into a method

Yes!, I think it is better.

**Review by obi-t4 — COMMENTED** [review]
I left some comments, but it seems there is no big issue I think.

File: app/ota_client.py (hunk 1 above)
Comment (Bodong-Yang) [reply app/ota_client.py:249]: Currently most important is dataclass will generate `__repr__` , `__eq__` and `__hash__`  for us automatically. 
`__repr__` is for debug purpose, `__eq__` and `__hash__` will allow customize set/dict behavior when storing Inf related class' instance(this feature will be used in later development).

File: app/ota_client.py (hunk 4 above)
Comment (Bodong-Yang) [reply app/ota_client.py:1201]: This replaces the previous implementation for logic of limiting concurrent __downloading__ tasks.

Comment (Bodong-Yang) [reply app/ota_client.py:1193]: The outer semaphore(which the `_collector` is maintaining), is used to limit the number of total concurrent tasks(download, copy, link).

Comment (obi-t4) [reply app/ota_client.py:1193]: > Is this not required?

Sorry, `if Ture:` is required?

Comment (Bodong-Yang) [reply app/ota_client.py:1193]: Oh I am sorry, I mistakenly left a debug used line, I will delete it now.

**Review by obi-t4 — APPROVED** [review]
LGTM :+1:

**Comment by Bodong-Yang** [comment]
> LGTM +1

Thanks! I will do a quick VM test for regression testing, after finish it I will merge the PR.

**Comment by Bodong-Yang** [comment]
Test on VM passed with `MAX_CONCURRENT_DOWNLOAD` set to 8 and `MAX_CONCURRENT_TASKS` bump to 1024!
